runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java

```diff
@@ -364,7 +366,12 @@ public void run() {
           } else if (update.getException().isPresent()) {
             visibleUpdates.offer(VisibleExecutorUpdate.fromThrowable(update.getException().get()));
           }
-          update = allUpdates.poll();
+          if (numUpdates >= MAX_UPDATES_PER_CALL) {
```

Comment:
Yes; processing an update is consists of submitting all of the consumer TransformExecutors (which do no work until invoked) to the underlying executor service.

However, I think time is probably the better thing to 'finish' based on, so I've switched to that.